src/common/oclengine.cpp

```diff
@@ -52,12 +55,34 @@ void OCLEngine::SetDeviceContextPtrVector(std::vector<DeviceContextPtr> vec, Dev
 
 void OCLEngine::SetDefaultDeviceContext(DeviceContextPtr dcp) { default_device_context = dcp; }
 
-OCLEngine::OCLEngine() { InitOCL(); }
+OCLEngine::OCLEngine()
+{
+    OCLInitResult res = InitOCL(false);
+    all_device_contexts = res.all_device_contexts;
+    default_device_context = res.default_device_context;
+}
 OCLEngine::OCLEngine(OCLEngine const&) {}
 OCLEngine& OCLEngine::operator=(OCLEngine const& rhs) { return *this; }
 
-void OCLEngine::InitOCL()
+OCLInitResult OCLEngine::InitOCL(bool saveBinaries)
 {
+#if defined(_WIN32) && !defined(__CYGWIN__)
```

Comment:
Pull this code out of this file into it’s own function.  Additionally, InitOCL should accept an explicit path to load the binary from - and main() should call the ‘get home directory’ code to pass that path in explicitly.  If no path is supplied, it should default to JIT only mode.  The main() should default to supplying a path unless `—no-cached-ocl` is passed in, which should always inhibit loading cached objects.

Reply:
Pulled it into its own static, public function. `InitOCL()` also now takes a path argument.

However, couple of problems. We don't supply `main()`, since we're a library to link against. The user always does, but we can give them the option to explicitly force JIT or cache. Also, in the absence of explicit overrides, I think the default behavior should be to try loading cached programs from the default location and fall back to JIT if they're absent or corrupt. This way, pre-existing Qrack programs don't need their code modified at all to turn program caching on and off; by default, it's up to presence/absence of valid cached programs in the default path.